mypy/types.py

```diff
@@ -144,6 +144,97 @@ def deserialize(cls, data: JsonDict) -> 'Type':
         raise NotImplementedError('Cannot deserialize {} instance'.format(cls.__name__))
 
 
+class TypeAliasType(Type):
+    """
+    A type alias to another type.
+
+    To support recursive type aliases we don't immediately expand a type alias
+    during semantic analysis, but create an in stance of this type that records the target alias
+    definition node (mypy.nodes.TypeAlias) and type arguments (for generic aliases).
+
+    This is very similar to how TypeInfo vs Instance interact.
+    """
+    def __init__(self, alias: Optional[mypy.nodes.TypeAlias], args: List[Type],
+                 line: int = -1, column: int = -1) -> None:
+        super().__init__(line, column)
+        self.alias = alias
+        self.args = args
+        self.type_ref = None  # type: Optional[str]
+
+    def expand_once(self) -> Type:
+        """Expand to the target type exactly once.
+
+        This doesn't do full expansion, i.e. the result can contain another
+        (or even this same) type alias.
+        """
+        assert self.alias is not None
+        return replace_alias_tvars(self.alias.target, self.alias.alias_tvars, self.args,
+                                   self.line, self.column)
+
+    def expand_all_if_possible(self) -> Optional['ProperType']:
+        """Attempt a full expansion of the type alias (including nested aliases).
+
+        If the expansion is not possible, i.e. the alias is (mutually-)recursive,
+        return None.
+        """
+        raise NotImplementedError('TODO')
+
+    @property
+    def can_be_true(self) -> bool:  # type: ignore
```

Comment:
The `# type: ignore` may break mypyc is strange ways, so it would be better not to use it.

Reply:
There is actually a mypy false positive https://github.com/python/mypy/issues/6759 that doesn't allow fixing this. I will leave a comment here with a link to issue.